Package delivery in mission (gripper / winch)

junwoo091400 opened the pull request:
## :bulb: About
![image](https://user-images.githubusercontent.com/23277211/181018274-f016f86c-d500-4890-a6ea-30d380d5a3c7.png)
Autonomous Package Delivery aims to provide an intuitive interface for users to have the drone deliver a package in Mission mode.

## :sob:  Current status of package delivery standard and it's limitations
There is a pre-existing “[MAV_CMD_NAV_PAYLOAD_PLACE](https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_PAYLOAD_PLACE)” message, which is supposed to handle the whole navigation to the specified waypoint, descend and release payload.

This message has 2 **flaws**:
1. The action depends on which vehicle type you have (For fixed wing, you can’t descend and release the payload smoothly), so implementing this item would involve a lot of customization for different vehicle types, which is **not generic**.
2. We are not specifying what kind of payload we are triggering and how, is it a gripper? a winch? If so, how should it behave? any timeouts? So it is **not specific enough**.

Therefore it was discussed that an ideal package delivery should achieve the following:
1. Do not use the Payload deploy message with Navigation feature (waypoint), but instead we would **support the specific Payload command (winch or gripper)**, which will only perform the action of deploying the payload.
2. Have a **timeout** so that the mission would wait until the timeout or acknowledgement of a successful payload deployment.

## :pretzel:  Proposed workflow
To use package deployment, a user should have a Mission plan file that has the following structure:
1. Previous mission item (Waypoint / Loiter / Land, etc)
2. Payload command (Gripper / Winch)

During execution time/timeout, the vehicle will be behaving according to the previous mission item, meaning if it was loitering, it would continue loitering, if it was a waypoint, it would stay at the waypoint, etc while delivering the payload.

This is a deliberate design decision to allow flexible usage of payload deployment (e.g. deploying while the airplane drone is loitering around the point of interest, as well as multicopter deploying the package while being landed).

### Payload protocols in MAVLink
There are 2 payload specific (no navigation involved) commands related to the delivery action available in MAVLink. This supports Gripper and Winch, but other payload types could be added as well. The message includes the following information:

1. [MAV_CMD_DO_GRIPPER](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_GRIPPER)
   * Gripper Instance number (Support for multiple grippers)
   * Gripper Action (Release / Grab)
2. [MAV_CMD_DO_WINCH](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_WINCH)
   * Winch Instance number (Support for multiple winches)
   * Winch Action (Freewheel the motor, Rate control the line, Abandon Line, etc.)
   * Winch Line Length [m] (Length of the winch to lower down)
   * Winch Lowering rate [m/s] (Rate at which winch will be lowered)

### Payload delivery timeout settings
The timeout settings are defined as parameters, for gripper application the settings are:

**Gripper Timeout (PD_GRIPPER_TO)**
Currently, as the feedback sensor isn’t supported gripper will never have the feedback on successful actuation. Therefore, this setting will be how long the payload deliverer will wait after which it will assume that the gripper actuation has been successful and send the successful vehicle_command_ack message to the navigator.

This setting is therefore hardware specific, and if you have a fast gripper, the timeout can be significantly lower.Proper timeout would be the time it takes for the gripper to go from closed position to open position. The default value is 3 seconds

**Mission Payload Delivery Timeout (MIS_PD_TO)**
This is a safeguard timeout setting for the mission, on how long the navigator will wait while waiting for a successful vehicle_command_ack of payload delivery. The default value is 5 seconds

## :building_construction: Architecture
![image](https://user-images.githubusercontent.com/23277211/181013866-edd476e3-28dd-4256-87b1-0b05fd87ab8d.png)

### Navigator
Navigator handles the reception of the vehicle command ack (described below). Upon reception of a successful deployment ack message, it sets the flag in Mission block level to signal that the payload deployment has been successful.
Vehicle Command Ack

We are waiting for the ack coming from either internally (via payload_deliverer module, which handles the Winch and Gripper actions), or externally (external entity sending the MAVLink message [COMMAND_ACK](https://mavlink.io/en/messages/common.html#COMMAND_ACK), which gets translated to vehicle_command_ack message in PX4).

### Mission
The Gripper / Winch command is placed as a ‘Mission Item’. This is possible since all the [Mission](https://mavlink.io/en/services/mission.html#mavlink_commands) Item has the MAV_CMD setting (e.g. Land, Takeoff, Waypoint, etc) which can get set to Gripper / Winch.

In the Mission logic (green box above) if either Gripper/Winch mission item is reached, it implements brake_for_hold functionality (which sets the valid flag of the next mission item waypoint to false) for rotary wings (e.g. Multicopter) so that the vehicle would hold it's position while the deployment is getting executed

For fixed wing and other platforms, no special braking condition is considered. So if you have a loiter mission item for a fixed-wing, it will execute the delivery whilst doing the loiter, and won’t come to a stop (as it’s impossible)

### Mission Block
MissionBlock is the parent class of Mission that handles the part “Is Mission completed?” is all performed in the is_mission_item_reached_or_completed function, to handle the time delay / mission item advancement.

Also it implements the actual issue_command function that sends out the vehicle command that activates the payload deployment.

### Payload Deliverer
This is a dedicated module that handles gripper / winch support.

## :test_tube: Test data / coverage
### Actual functional test of Landed package delivery (06.07.2022)
This test shows how a landed vehicle actuates the gripper to drop the package at the timestamp marked with the red line below, then takes off and resumes the mission.

![image](https://user-images.githubusercontent.com/23277211/181019930-9b8f981c-c8ed-414b-bc70-b7e3fa346733.png)

The log can be found here: https://logs.px4.io/plot_app?log=3e8728f2-00df-4807-b1c5-d8cf1d2afec6

### SITL test
The mission plan file for SITL test is here (change the extension to .plan to import into QGC): [gripper_mission_sitl.txt](https://github.com/PX4/PX4-Autopilot/files/9190227/gripper_mission_sitl.txt)

## :book: How to use Package delivery feature (for documentation)
### Enable the payload_deliverer module
Do the `make <your-target-name> boardconfig` and **include** the `payload_deliverer` module using the Kconfig editor.

### Parameter setup
Configure the Gripper:
1. Set **PD_GRIPPER_EN** to **Enabled** : Enables actuating the gripper
1. Set **PD_GRIPPER_TO** to the timeout [s] of how long the gripper opening actuation will take
> For example, if you command the gripper to open and it takes ~3 seconds to fully open, set it to 3.0

Configure the Navigator:
* Set **MIS_PD_TO** to the timeout [s] you want in Navigator to wait for the successful payload delivery acknowledgement
> This is the safeguard against the payload delivery never sending back the acknowledgement and mission being in hold for infinite amount of time.

Other parameters:
* Set **COM_PREARM_MODE** to **Always**, this enalbwes the payload deployment mechanism to work (since it’s not a Motor) without arming (for putting in the payload at the bench).
> Without setting this, your actuator will be stuck in ‘Disarmed' position (usually PWM command 900, at a retracted position), and won’t work unless you are armed!

## :monkey: Additional context
#19659 has made the default behavior of having a land item in the middle of a mission to not disarm but to keep in idle state. Package delivery also relies on the fact that mission wouldn't get automatically terminated during land, and this issue has been addressed here: https://github.com/mavlink/mavlink/pull/1867

bkueng commented on src/modules/payload_deliverer/CMakeLists.txt:
```diff
@@ -0,0 +1,45 @@
+	MODULE_CONFIG
+		module.yaml
+	COMPILE_FLAGS
+		-DDEBUG_BUILD
```
```suggestion
```

bkueng commented on src/modules/payload_deliverer/gripper.cpp:
```diff
@@ -0,0 +1,137 @@
+
+Gripper::Gripper()
+{
+	// Do nothing
```
You can use `Gripper() = default;`
```suggestion
```

bkueng commented on src/modules/payload_deliverer/gripper.h:
```diff
@@ -0,0 +1,129 @@
+	GripperType type{GripperType::UNDEFINED};
+
+	// Gripper state feedback sensor type
+	typedef enum GripperSensorType {
+		NONE = -1,
+		ENCODER,
+	} GripperSensorType;
```
```suggestion
	enum class GripperSensorType {
		NONE = -1,
		ENCODER,
	};
```

junwoo091400 replied:
Done! What would be a big difference between typedef enum vs enum class?

bkueng replied:
`enum class` is the modern variant and has a few benefits.

bkueng commented on src/modules/payload_deliverer/payload_deliverer.cpp:
```diff
@@ -0,0 +1,255 @@
+		return false;
+
+	} else {
+		_gripper.grab(); // Initialize to grab position
```
This is already done in the gripper init. I would remove it there.

junwoo091400 replied:
Makes sense!

bkueng commented on src/modules/payload_deliverer/payload_deliverer.cpp:
```diff
@@ -0,0 +1,255 @@
+
+void PayloadDeliverer::parameter_update()
+{
+	// Call the ModuleParams function to update internal parameters
```
Unnecessarily verbose
```suggestion
```

bkueng commented on src/modules/payload_deliverer/payload_deliverer.cpp:
```diff
@@ -0,0 +1,255 @@
+void PayloadDeliverer::Run()
+{
+	if (should_exit()) {
+		ScheduleClear();
```
You need to unregister the callback

bkueng commented on src/modules/payload_deliverer/payload_deliverer.cpp:
```diff
@@ -0,0 +1,255 @@
+
+void PayloadDeliverer::update_gripper(const hrt_abstime &now,  const vehicle_command_s *vehicle_command)
+{
+	if (!_param_gripper_enable.get()) {
+		PX4_WARN("Gripper isn't enabled, so the gripper functionality will not work!");
+		// If gripper isn't enabled, don't do anything
+		return;
+	}
```
Why check for this, the module isn't expected to run if disabled.

junwoo091400 replied:
Good point. I instead changed to check if gripper instance is valid (since that actually can happen during runtime, if gripper setting isn't correct)

bkueng commented on src/modules/payload_deliverer/payload_deliverer.cpp:
```diff
@@ -0,0 +1,255 @@
+	}
+
+	// Process if we received DO_GRIPPER vehicle command
+	if (vehicle_command -> command == vehicle_command_s::VEHICLE_CMD_DO_GRIPPER) {
```
```suggestion
	if (vehicle_command->command == vehicle_command_s::VEHICLE_CMD_DO_GRIPPER) {
```

bkueng commented on src/modules/payload_deliverer/payload_deliverer.cpp:
```diff
@@ -0,0 +1,255 @@
+	_gripper.release();
+	px4_usleep(5_s);
+	PX4_DEBUG("Test: Closing the Gripper!");
+	_gripper.grab();
```
Better publish a vehicle command

junwoo091400 replied:
But that wouldn't work since the `gripper_test` function would be blocking (the `Run()` execution would come to halt) while it is in use, and hence the vehicle command won't get registered and actuate the gripper (which is done in `Run()`).

I have just tested this by activating the command, and I can verify that actuator doesn't move (due to the reason given above, I assume).

Or am I missing something?

junwoo091400 replied:
![image](https://user-images.githubusercontent.com/23277211/183675407-b2febff6-d5ab-4352-a092-f2d6a9387d82.png)

As shown here, using the vehicle command, it doesn't get translated into the gripper command as expected while executing the gripper_test.

So I think sticking to utilizing `_gripper` object for grab/releasing is ok, what do you think?

bkueng replied:
`custom_command` is executed on a different thread so it should work.

junwoo091400 replied:
I will try again, but it seemed like the program was on hold while the gripper_test() function was executing 🤔

junwoo091400 replied:
Solved the problem in the latest commit here: https://github.com/PX4/PX4-Autopilot/pull/19963/commits/9e73973dc4ccc982df9ae408e4f4d91dd7a5a541

There was a bug on my side, due to the SubscriptionCallback usage. Details are explained in the commit.

junwoo091400 commented on msg/vehicle_command.msg:
```diff
@@ -149,30 +185,9 @@ uint8 FAILURE_TYPE_SLOW = 5
+# Queue length for the uORB topic instance
 uint8 ORB_QUEUE_LENGTH = 8
 
-float32 param1			# Parameter 1, as defined by MAVLink uint16 VEHICLE_CMD enum.
```
I grouped the variables back together :+1:

junwoo091400 commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -128,11 +131,13 @@ void Navigator::params_update()
 void Navigator::run()
 {
-	bool have_geofence_position_data = false;
+	_have_geofence_position_data = false;
```
Reverted the changes. I at first thought it was getting used too much to be a local variable, but it doesn't need to be :wink:

junwoo091400 commented on src/modules/navigator/navigator.h:
```diff
@@ -331,17 +339,17 @@ class Navigator : public ModuleBase<Navigator>, public ModuleParams
 	uORB::Subscription _land_detected_sub{ORB_ID(vehicle_land_detected)};	/**< vehicle land detected subscription */
 	uORB::Subscription _pos_ctrl_landing_status_sub{ORB_ID(position_controller_landing_status)};	/**< position controller landing status subscription */
 	uORB::Subscription _traffic_sub{ORB_ID(transponder_report)};		/**< traffic subscription */
-	uORB::Subscription _vehicle_command_sub{ORB_ID(vehicle_command)};	/**< vehicle commands (onboard and offboard) */
```
No reason except that it would reduce the verbose allocation of a `vehicle_command_s` struct object in the function. Would this have a drawback?

bkueng replied:
Yes, the class member struct uses more memory.

junwoo091400 replied:
Applied in the latest commit! https://github.com/PX4/PX4-Autopilot/pull/19963/commits/9d5f2d8916a9b01d48063c915636e213b6f2891e

junwoo091400 commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -217,15 +222,19 @@ void Navigator::run()
 				PX4_ERR("vehicle_command lost, generation %d -> %d", last_generation, _vehicle_command_sub.get_last_generation());
 			}
 
+			// Partial solution to keep 'cmd' as local variable to not create such a big diff in the PR
+			// TODO: Put all of this below into a function named: ""
```
Done!

junwoo091400 commented:
Rebased on `main` and resolved conflicts regarding the FucntionGripper header location change that occurred.

junwoo091400 commented:
Note on what I learned during rebase: Without setting `COM_PREARM_MODE` to `Always`, I naively thought the gripper control was not working, but this was necessary for testing the Gripper functionality :facepalm:

bkueng commented on src/lib/mixer_module/functions.hpp:
```diff
@@ -0,0 +1,313 @@
+ *
+ ****************************************************************************/
+
+#pragma once
```
Something went wrong here during the rebase - this got split into several files now.

junwoo091400 replied:
Thanks!

junwoo091400 replied:
Rebase fixed and now the files are removed!

junwoo091400 commented:
Rebased on `main`, removed unnecessary mixer based payload delivery actuation airframe file modifications as well.

junwoo091400 commented:
1. Squashed into a singular commit
2. Removed unnecessary files
3. Removed unnecessary changes (whitespaces / formatting) in uORB message definitions (vehicle_command)

hamishwillee commented:
Interesting. I think I would have used a [MAV_CMD_NAV_LOITER_TIME](https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_LOITER_TIME) or [MAV_CMD_NAV_LOITER_UNLIM](https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_LOITER_UNLIM) to provide the timeout in a tunable way. A landed vehicle would loiter by waiting in place.
